lib/storage/src/content_manager/errors.rs

```diff
@@ -15,8 +17,60 @@ pub enum StorageError {
     ServiceError { description: String },
     #[error("Bad request: {description}")]
     BadRequest { description: String },
+    #[error("IO error: {description}")]
+    IO {
+        description: String,
+        source: Option<Rc<anyhow::Error>>,
+    },
+    #[error("parsing error: {description}")]
+    Parse {
+        description: String,
+        source: Option<Rc<anyhow::Error>>,
+    },
+}
+impl StorageError {
+    pub(crate) fn from_std_io_error_with_msg(
+        msg: impl Into<String>,
+        source: std::io::Error,
+    ) -> Self {
+        StorageError::IO {
+            description: msg.into(),
+            source: Some(Rc::new(source.into())),
+        }
+    }
+    pub(crate) fn from_any_io_error_with_msg(
+        msg: impl Into<String>,
+        source: anyhow::Error,
+    ) -> Self {
+        StorageError::IO {
+            description: msg.into(),
+            source: Some(Rc::new(source)),
+        }
+    }
+    pub(crate) fn from_utf8_error_with_msg(msg: impl Into<String>, source: Utf8Error) -> Self {
+        StorageError::Parse {
+            description: msg.into(),
+            source: Some(Rc::new(source.into())),
+        }
+    }
+    pub(crate) fn from_io_msg(msg: impl Into<String>) -> Self {
+        StorageError::IO {
+            description: msg.into(),
+            source: None,
+        }
+    }
```

Comment:
`impl From<(ToString, Utf8Error)> for StorageError` might be more like rust-way approach, what do you think?

Overall I would prefer to minimize error-handling boilerplate code in the functions and make use of `?` as much as possible. Maybe it is possible to do something with `Context` - https://rust-cli.github.io/book/tutorial/errors.html#providing-context ?

Reply:
good point.

Reply:
I usually use From only to convert from a thing taking one arg but the tuple can be a good candidate too.